Add support for nested user-name-attribute using dot notation

yybmion opened the pull request:
Closes gh-16390

Hi team, this PR adds support for nested properties in the OAuth2 user-name-attribute using dot notation.
For example, if a provider returns a nested response like `{"data": {"username": "user1"}}`,
users can now directly configure `user-name-attribute: data.username` without writing custom code.

## Implementation
- Added extractNestedAttribute utility method to extract values from nested paths
- Modified loadUser method to handle attributes with dot notation
- Applied the same changes to both servlet and reactive implementations
- Added comprehensive tests to verify functionality

## Testing
Added tests for
- Simple nested attributes 
- Multi-level nested attributes 

All tests pass and existing functionality is preserved.

rwinch commented:
Thanks for the PR @yybmion! I wonder if this might be better implemented using SpEL to provide more powerful options for resolving the username. What are your thoughts?

rwinch commented:
Yes. Please provide an implementation that uses SpEL.

yybmion commented:
Hello @rwinch, I'd like to clarify your feedback on my PR about supporting nested properties in the user-name-attribute.

Did you mean that I should implement support for expressions like `#{data.username}` in the properties and yml configuration files to handle nested structures? (Instead using `data.username`)

Thank you for your guidance!

rwinch commented:
I think an outline would be:

## Allow Injecting the Principal Name into `DefaultOAuth2User`

Right now `OAuth2UserService` requires defining the `name` property as an attribute name. This limits the flexibility of resolving the name.

Update `DefaultOAuth2User` to allow injecting the name (rather than a property for obtaining the name). You would add a new member variable of `username`. This would likely require using a static factory method to distinguish from the existing constructor since the types of the `nameAttributeKey` and `username` are the same.

You can remove the `nameAttributeKey` member variable and instead populate the `username` using the `attributes[nameAttributeKey]` in the constructor if `nameAttributeKey` was specified.

Deprecate the old `DefaultOAuth2User` constructor in favor of injecting the name directly and update the usage of `DefaultOAuth2User` within Spring Security to no longer use the deprecated constructor.

## Add SpEL Support

Update `ClientRegistration` to have a property named `usernameExpression` and  remove the `userNameAttributeName` property but preserving and deprecating the methods which instead populate the `usernameExpression` member variable. This should be passive since the `usernameAttributeName` is a valid SpEL expression. 

Update `OAuth2UserService` to extract the username using the `usernameExpression` property from the `ClientRegistration` as a SpEL expression with the attributes as the root object. Create the `DefaultOAuth2User` with the username injected into it rather than using the `nameAttributeKey`.

I think creating these as two separate commits is valuable since they are useful on their own. Let me know your thoughts.

rwinch commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -552,17 +572,25 @@ public Builder userInfoAuthenticationMethod(AuthenticationMethod userInfoAuthent
+		 * @deprecated Use {@link #usernameExpression(String)} instead
 		 */
+		@Deprecated
 		public Builder userNameAttributeName(String userNameAttributeName) {
```
This is the place where I think you should default the usernameExpression to `"['" + userNameAttributeName + `"']"`.

rwinch commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -552,17 +572,25 @@ public Builder userInfoAuthenticationMethod(AuthenticationMethod userInfoAuthent
-		 * @return the {@link Builder}
+		 * @deprecated Use {@link #usernameExpression(String)} instead
 		 */
+		@Deprecated
```
I think that you can remove the Deprecated as this simplifies setting the expression. Just update the Javadoc.

rwinch commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -552,17 +572,25 @@ public Builder userInfoAuthenticationMethod(AuthenticationMethod userInfoAuthent
 		}
 
+		/**
+		 * Sets the SpEL expression used to extract the username from user info response.
```
Please provide a few examples for the value.

rwinch commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:
```diff
@@ -77,11 +82,47 @@ public DefaultOAuth2User(Collection<? extends GrantedAuthority> authorities, Map
+	 * @param username the user's name
+	 * @return a new {@code DefaultOAuth2User}
+	 */
+	public static DefaultOAuth2User withUsername(Collection<? extends GrantedAuthority> authorities,
```
I think mentioned a factory method previously, but I do not like this with multiple arguments. Please create a `public static Builder withUsername(String)` method instead.

rwinch reviewed (changes requested):
Thank you for the updates. This is taking shape. I've provided feedback inline.

yybmion commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:
```diff
@@ -90,13 +97,69 @@ public DefaultOAuth2UserService() {
+		}
+		else {
+			try {
+				ExpressionParser parser = new SpelExpressionParser();
```
Done.

yybmion commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:
```diff
@@ -90,13 +97,69 @@ public DefaultOAuth2UserService() {
-		return new DefaultOAuth2User(authorities, attributes, userNameAttributeName);
+
+		String evaluatedUsername = evaluateUsername(attributes, usernameExpression);
+		Collection<GrantedAuthority> authorities = getAuthorities(token, attributes, usernameExpression);
```
Thanks for highlighting this important integration issue with gh-15012!

I've addressed this by following your suggestion to expose the username directly rather than the attribute name.

- Added `username` field to OAuth2UserAuthority with `getUsername()` method
- Introduced Builder pattern: `OAuth2UserAuthority.withUsername(evaluatedUsername)`
- Pass the evaluated username result (not the SpEL expression) to OAuth2UserAuthority

I think this way, OAuth2UserAuthority always receives the final evaluated username string, preserving compatibility with gh-15012 regardless of expression complexity.

yybmion commented:
Thanks for the feedback @rwinch ! I've addressed all the review comments and updated the code accordingly.

rwinch commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/OAuth2UserAuthority.java:
```diff
@@ -44,6 +45,8 @@ public class OAuth2UserAuthority implements GrantedAuthority {
 
 	private final String userNameAttributeName;
```
This property should be deprecated as should all the methods/constructors that use it.

Please also address `OidcUserAuthority` which extends from this class. It will likely need it's own `Builder` that extends from this `Builder` and have an overridden `withUsername` static method.

rwinch commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/OAuth2UserAuthority.java:
```diff
@@ -88,6 +91,39 @@ public OAuth2UserAuthority(String authority, Map<String, Object> attributes, Str
+	 * the attributes - preserved for backwards compatibility
+	 * @param username the username
+	 */
+	private OAuth2UserAuthority(String authority, Map<String, Object> attributes, String userNameAttributeName,
```
since the Builder doesn't use `usernameAttributeName`, we shouldn't include this as a property here, so that this constructor does not deprecated properties.

rwinch replied:
Consider the order of username, authority, attributes since username is now a required property.

rwinch commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/OAuth2UserAuthority.java:
```diff
@@ -113,6 +149,18 @@ public String getUserNameAttributeName() {
+	 * This method provides direct access to the username without requiring knowledge of
+	 * the attribute structure or SpEL expressions used to extract it.
+	 * @return the username
+	 * @since 6.5
```
The PR targets main, so this (and any other since) should be since 7.0

rwinch reviewed (changes requested):
This is getting very close!

Please address the comments and ensure to update the documentation (including the What's New Section)

yybmion commented:
Hi @rwinch.  I've addressed all the review feedback and updated the documentation as requested. 
Please let me know if there are any areas in the documentation that need further improvement or clarification.

rwinch commented on docs/modules/ROOT/pages/reactive/oauth2/client/core.adoc:
```diff
@@ -36,12 +36,12 @@ public final class ClientRegistration {
 			private AuthenticationMethod authenticationMethod;  <15>
 			private String userNameAttributeName;	<16>
-
+            private String usernameExpression;    <17>
```
Can you please fix this to use tabs?

rwinch commented on docs/modules/ROOT/pages/reactive/oauth2/login/core.adoc:
```diff
@@ -144,6 +144,9 @@ The following table outlines the mapping of the Spring Boot OAuth Client propert
 |`providerDetails.userInfoEndpoint.userNameAttributeName`
+
+|`spring.security.oauth2.client.provider._[providerId]_.username-expression`
+|`providerDetails.userInfoEndpoint.usernameExpression`
```
This won't work yet because Spring Boot has not added support. Please remove it for now. If you like, you can save it in another branch in its own commit and we can use it when Spring Boot adds the support.

rwinch commented on docs/modules/ROOT/pages/servlet/oauth2/login/core.adoc:
```diff
@@ -142,6 +142,9 @@ The following table outlines the mapping of the Spring Boot OAuth Client propert
 |`providerDetails.userInfoEndpoint.userNameAttributeName`
+
+|`spring.security.oauth2.client.provider._[providerId]_.username-expression`
+|`providerDetails.userInfoEndpoint.usernameExpression`
```
This won't work yet because Spring Boot has not added support. Please remove it for now. If you like, you can save it in another branch in its own commit and we can use it when Spring Boot adds the support.

rwinch commented on docs/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -8,3 +8,29 @@ Below are the highlights of the release, or you can view https://github.com/spri
+
+OAuth2 Client now supports SpEL expressions for extracting usernames from nested UserInfo responses, eliminating the need for custom `OAuth2UserService` implementations in many cases. This is particularly useful for APIs like Twitter API v2 that return nested user data:
+
+[source,yaml]
```
The yaml won't work yet because Spring Boot has not added support. Please remove it for now. If you like, you can save it in another branch in its own commit and we can use it when Spring Boot adds the support.

rwinch reviewed (changes requested):
Thank you for the changes.

I've included some comments inline around Spring Boot not supporting the new property yet. We will need to create a ticket with them once we merge here. Once they add the support we can add the documentation back.

Please go through and find/remove any usage (within main) for any of the constructors/methods that you deprecated. For example, I see that `OidcUserRequestUtils` still references `UserInfoEndpoint.getUserNameAttributeName()` We should update it to use non-deprecated methods.

yybmion commented:
Thank you for the review @rwinch ! 
I'm planning to create a separate commit to address the deprecated constructor/method usage you mentioned. However, I have a few questions that came up during the modifications.

I'm wondering about your thoughts on :

1. Deprecating the nameAttributeKey field in DefaultOAuth2User constructors for consistency with the new usernameExpression approach
2. Adding builder pattern to DefaultOidcUser since it extends DefaultOAuth2User and currently calls the deprecated parent constructor via super()

Should I include these changes in this PR?

rwinch commented:
Thanks for reaching out and your patience as we put the final touches on this @yybmion 

> Deprecating the nameAttributeKey field in DefaultOAuth2User constructors for consistency with the new usernameExpression approach

Yes. I think that this is good. Generally, we want people to move to the builder instead.

> Adding builder pattern to DefaultOidcUser since it extends DefaultOAuth2User and currently calls the deprecated parent constructor via super().

Yes I think that is good too. You will likely need to ensure the non-deprecated constructor in DefaultOAuth2User is protected so that it can be used by DefaultOidcUser.

> Should I include these changes in this PR?

Yes please. It is important that we push people towards the new, more powerful, approach.

yybmion commented:
Hi @rwinch, I noticed that the "What's New" documentation for Spring Security 7 has recently been updated, which caused a conflict with the changes I previously added. I've resolved the conflict accordingly.

yybmion commented:
Hi @rwinch,
I noticed that the "What's New" documentation for Spring Security 7 has recently been updated, which caused a conflict with the changes I previously added. I've resolved the conflict accordingly.

rwinch commented:
Thanks again for all of your work on this PR!

I'd like to encapsulate the logic (keep private) in `OAuth2UsernameExpressionUtils` so that we can refactor without breaking anything later. One example of such a refactoring is that at some point the `ExpressionParser` might need to be injected to allow resolving Bean references in the expressions. Even if we allow injecting the `ExpressionParser`, I'd still like to keep this logic encapsulated in case of other refactorings.

To work toward that end I created and resolved gh-17626. In the case where we invoke the `OAuth2UserService` (or the reactive equivalent), we could use the result to obtain the username. The problem is that OidcUserService (and the reactive equivalent) leverage  `userNameAttributeName` even when the `OAuth2Service` is not invoked and it does not use the result from the delegate service even when it does invoke the user info endpoint.

To address those concerns I created gh-17627 gh-17628

I believe we could extract the logic for injecting the username rather than the attribute into a separate PR and merge that without being blocked since that behavior would be passive and not be impacted by the decisions of the issues above. Feel free to do this if you like. However, as of now we are blocked on merging resolving the username using an expression until those two issues are resolved.

cc @jgrandja

rwinch commented:
FYI I rebased and pushed the commits

yybmion commented:
Thanks for the rebase and I understood about the blocking issues.
